Mathlib/Tactic/Change.lean

```diff
@@ -0,0 +1,39 @@
+/-
+Copyright (c) 2023 Damiano Testa. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Damiano Testa
+-/
+
+import Mathlib.Tactic.Basic
+/-!
+
+# Tactic `change? _`
+
+This tactic is used to replace the goal with a definitionally equal one.
+The `_` is a general term and is intended to be used to guide which unfolding you want
+Lean to perform.
+-/
+
+/--  `change? term?` unifies the optional `term?` with the current goal, defaulting to the
+goal, if `term?` is not present.
+It then prints in the infoView the unified term.
+This is useful to replace the main goal after a `dsimp`.
+```lean
+example : (fun x : Nat => x) 0 = 1 := by
+  change? 0 = _  -- `change 0 = 1`
+```
+-/
+syntax (name := change?) "change?" (ppSpace colGt term)? : tactic
+
+open Lean Meta Elab.Tactic Std.Tactic.TryThis in
+@[inherit_doc change?, tactic change?] def elabChange? : Tactic := fun stx => do
+  let sop := stx[1].getOptional?
+  let expr := ← match sop with
+    | none => getMainTarget
+    | some ex => do
+      let ex := ← elabTerm ex none
+      let defeq? := ← isDefEq ex (← getMainTarget)
+      if ! defeq? then throwError "The given term is not DefEq to the goal"
+      instantiateMVars ex
+  let dstx := ← delabToRefinableSyntax expr
+  addSuggestion stx (← `(tactic| change $dstx))
```

Comment:
For ease of maintenance by others, this should probably use `elab_rules` syntax instead of low-level syntax manipulation. Also, it seems to be standard to put the blue squiggle under the initial token instead of the whole thing; we pass `origSpan?` to tell `addSuggestion` to replace the whole thing upon clicking, though. :)
```suggestion
elab_rules : tactic
| `(tactic|change?%$tk $[$sop:term]?) => do
  let stx ← getRef
  let expr := ← match sop with
    | none => getMainTarget
    | some ex => do
      let ex := ← elabTerm ex (← inferType <|← getMainTarget)
      let defeq? := ← isDefEq ex (← getMainTarget)
      if ! defeq? then throwError "The given term is not DefEq to the goal"
      instantiateMVars ex
  let dstx := ← delabToRefinableSyntax expr
  addSuggestion tk (← `(tactic| change $dstx)) (origSpan? := stx)
```

Reply:
I didn't know about `getRef`, which is why I had used the weird definition.  It was my hacky way of getting my hands on the whole syntax.

Reply:
Ah, gotcha—I thought that might have been why. It took a second for me to locate it as well.